zproject/backends.py

```diff
@@ -414,7 +414,7 @@ class ZulipLDAPAuthBackend(ZulipLDAPAuthBackendBase):
     def authenticate(self, username, password, realm_subdomain=None, return_data=None):
         # type: (Text, str, Optional[Text], Optional[Dict[str, Any]]) -> Optional[UserProfile]
         try:
-            if settings.REALMS_HAVE_SUBDOMAINS:
+            if settings.REALMS_HAVE_SUBDOMAINS and realm_subdomain is not None:
```

Comment:
This certainly makes me appreciate the tests!

After another look, perhaps the issue is simply that the function depends on `settings` to know whether to just use `get_realm`, which is an independent factor from the `realm_subdomain` parameter. On that basis, perhaps the right approach might be to move the code into the first branch?
```
if settings.REALMS_HAVE_SUBDOMAINS:
    if realm_subdomain is None:
        raise SomeException("REALMS_HAVE_SUBDOMAINS set, but no subdomain supplied")  # or return None?
    self._realm = get_realm(realm_subdomain)
else:
    self._realm = get_realm_by_email_domain(username)
```